object/service.go

```diff
@@ -431,3 +432,82 @@ func (s *Service) DeleteObjects(ctx aws.Context, bucket *string, keys []*string)
 	}
 	return deleteResponse, nil
 }
+
+
+// CopyObject - copy an object between source and destination buckets
+// It receives a source bucket, object key and a destination bucket 
+func (s *Service) CopyObject(
+	ctx aws.Context,
+	bucketSrc *string, 
+	bucketDest *string, 
+	keySrc *string,  
+	keyDest *string,
+	) (*string, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("context is required")
+	}
+
+	if bucketSrc == nil || *bucketSrc == "" {
+		return nil, fmt.Errorf("source bucket name is required")
+	}
+
+	if bucketDest == nil || *bucketDest == "" {
+		return nil, fmt.Errorf("destination bucket name is required")
+	}
+
+	if keySrc == nil || *keySrc == "" {
+		return nil, fmt.Errorf("object's src key is required")
+	}
+
+	if keyDest == nil || *keyDest == "" {
+		return nil, fmt.Errorf("object's dest key is required")
+	}
+
+	
+	// Check if the source bucket exists 
+	// if it doesn't exist, we don't need to create a new bucket,
+	// because it would be empty with no object to copy
+	headBucketinput := &s3.HeadBucketInput{Bucket: bucketSrc}
+
+	if _, err := s.s3Client.HeadBucket(headBucketinput); err != nil {
+		return nil, fmt.Errorf("failed to CopyObject from bucket, %s, does not exist: %v", *bucketSrc, err)
+	}
+
+	// Check if the object exists
+	sourceObjectResponse, err := s.HeadObject(ctx, keySrc, bucketSrc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to CopyObject from bucket, %s, because object %s does not exist: %v", *bucketSrc, *keySrc, err)
+	}
+
+	// Check if the destination bucket exist 
+	if err := s.ensureBucketExists(ctx, bucketDest); err != nil {
+		return nil, fmt.Errorf("failed to CopyObject from bucket, %s, does not exist: %v", *bucketSrc, err)
+	}
+
+	// Parse the location of the object to URL
+	objectToCopy := url.QueryEscape(*bucketSrc + "/" + *keySrc)
+
+	copyObjectinput := &s3.CopyObjectInput{
+		Bucket: bucketDest,
+		CopySource: aws.String(objectToCopy), // (CopySource field expected url)
+		Key: keyDest,
+	}
+
+	copyObjectResponse, err := s.s3Client.CopyObjectWithContext(ctx, copyObjectinput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to copy object: %v", err)
+	}
+
+	// Compare the ETags between the source and the destination buckets (kind of a checksum)
+	if *sourceObjectResponse.ETag != *copyObjectResponse.CopyObjectResult.ETag {
+		return nil, fmt.Errorf(
+			"failed to copy object %s from source bucket, %s, because something went wrong in the process of copying the object to bucket %s and the ETag has changed : %v",
```

Comment:
```suggestion
			"failed to copy object %s from source bucket, %s, to destination bucket %s. ETag has changed : %v",
```
The error message is too long

Reply:
done